policies/templates/gcp_restricted_firewall_rules_v1.yaml

```diff
@@ -0,0 +1,628 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# This template is for policies restricting which firewall rules
+# should exist in a specific location in GCP.
+# All firewall rules matching the passed parameters will be flagged as violations.
+# All parameters are optional.
+#
+# A firewall rule will trigger a violation ONLY IF it matches ALL the passed parameters.
+
+apiVersion: templates.gatekeeper.sh/v1alpha1
+kind: ConstraintTemplate
+metadata:
+  name: gcp-restricted-firewall-rules-v1
+spec:
+  crd:
+    spec:
+      names:
+        kind: GCPRestrictedFirewallRulesConstraintV1
+        plural: gcprestrictedfirewallrulesconstraintsv1
+      validation:
+        openAPIV3Schema:
+          properties:
+            rules:
+              description: "List of rules to match in order to raise violations. Rule are object describing rules (see below)"
+              type: array
+              items: objects
+              additionalProperties:
+                direction:
+                  description: "Direction of the rules to restrict."
+                  type: string
+                  enum: ["INGRESS", "EGRESS", "any"]
+                  default: "any"
+                rule_type:
+                  description: "Type of the rules to restrict."
+                  type: string
+                  enum: ["allowed", "denied", "any"]
+                  default: "any"
+                port:
+                  description: |
+                    "Port of the rules to restrict. This parameter is tighly coupled with the
+                    protocol paramter: the rule will be marked as in violation if the parameter
+                    protocol:port pair matches one of its
+                    protocol:port listed.
+                    Valid values are:
+                        * single number
+                        * ranges (e.g 22-25) - Note: only FW rules with a range greater or equal
```

Comment:
I think logic should be that any rule within the provided range would be a violation?

Reply:
The logic of this policy is to let you describe fw rules that should raise a violation. So the way the range parameter works is that if a fw rule range includes the passed range (here) then it raises a violation.

My intention was that, as a user I want to flag all fw rules that match specific criteria, so here I considered that if I want to flag rules with ports 20-30 and and a fw happens to be 10-40, I still flag it since my criteria is true for that rule (it includes 20-30). 

Is it something we should change? The original PRD was only exact matches for ports, even for ranges.

Reply:
Good point, that makes sense.